consensus/hotstuff/timeoutcollector/timeout_collector_test.go

```diff
@@ -0,0 +1,186 @@
+
+import (
+	"errors"
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/consensus/hotstuff/helper"
+	"github.com/onflow/flow-go/consensus/hotstuff/mocks"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+func TestTimeoutCollector(t *testing.T) {
+	suite.Run(t, new(TimeoutCollectorTestSuite))
+}
+
+// TimeoutCollectorTestSuite is a test suite for testing TimeoutCollector. It stores mocked
+// state internally for testing behavior.
+type TimeoutCollectorTestSuite struct {
+	suite.Suite
+
+	view              uint64
+	notifier          *mocks.Consumer
+	collectorNotifier *mocks.TimeoutCollectorConsumer
+	processor         *mocks.TimeoutProcessor
+	collector         *TimeoutCollector
+}
+
+func (s *TimeoutCollectorTestSuite) SetupTest() {
+	s.view = 1000
+	s.notifier = mocks.NewConsumer(s.T())
+	s.collectorNotifier = mocks.NewTimeoutCollectorConsumer(s.T())
+	s.processor = mocks.NewTimeoutProcessor(s.T())
+
+	s.collectorNotifier.On("OnNewQcDiscovered", mock.Anything).Maybe()
+	s.collectorNotifier.On("OnNewTcDiscovered", mock.Anything).Maybe()
+
+	s.collector = NewTimeoutCollector(s.view, s.notifier, s.collectorNotifier, s.processor)
+}
+
+// TestView tests that `View` returns the same value that was passed in constructor
+func (s *TimeoutCollectorTestSuite) TestView() {
+	require.Equal(s.T(), s.view, s.collector.View())
+}
+
+// TestAddTimeout_HappyPath tests that process in happy path executed by multiple workers deliver expected results
+// all operations should be successful, no errors expected
+func (s *TimeoutCollectorTestSuite) TestAddTimeout_HappyPath() {
+	var wg sync.WaitGroup
+	for i := 0; i < 20; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			timeout := helper.TimeoutObjectFixture(helper.WithTimeoutObjectView(s.view))
+			s.processor.On("Process", timeout).Return(nil).Once()
+			err := s.collector.AddTimeout(timeout)
+			require.NoError(s.T(), err)
+		}()
+	}
+
+	wg.Wait()
+	s.processor.AssertExpectations(s.T())
+}
+
+// TestAddTimeout_DoubleTimeout tests that submitting two different timeouts for same view ends with reporting
+// double timeout to notifier which can be slashed later.
+func (s *TimeoutCollectorTestSuite) TestAddTimeout_DoubleTimeout() {
+	timeout := helper.TimeoutObjectFixture(helper.WithTimeoutObjectView(s.view))
+	s.processor.On("Process", timeout).Return(nil).Once()
+	err := s.collector.AddTimeout(timeout)
+	require.NoError(s.T(), err)
+
+	otherTimeout := helper.TimeoutObjectFixture(helper.WithTimeoutObjectView(s.view),
+		helper.WithTimeoutObjectSignerID(timeout.SignerID))
+
+	s.notifier.On("OnDoubleTimeoutDetected", timeout, otherTimeout).Once()
+
+	err = s.collector.AddTimeout(otherTimeout)
+	require.NoError(s.T(), err)
+	s.notifier.AssertCalled(s.T(), "OnDoubleTimeoutDetected", timeout, otherTimeout)
```

Comment:
```suggestion
	s.notifier.AssertExpectations(s.T())
```
nit: can use `AssertExpectations` to avoid duplicating the `On` call